bluepysnap/nodes.py

```diff
@@ -32,12 +32,201 @@
+                - ``None``: return all node IDs of the circuit in a CircuitNodeIds object.
+
+        Returns:
+            CircuitNodeIds: returns a CircuitNodeIds containing all the node IDs and the
+                corresponding populations. All the explicitly requested IDs must be present inside
+                the circuit.
+
+        Raises:
+            BluepySnapError: if a population is explicitly required by the user (via CircuitNodeIds)
+            BluepySnapError: if one or more ids are explicilty asked by the user using int,
+                sequence, or CircuitNodeIds.
+
+        Examples:
+            The available group parameter values (example with 2 nodes populations pop1 and pop2):
+            >>> nodes.ids(group=None)  #  returns all CircuitNodeIds from the circuit
+            >>> node_ids = CircuitNodeIds.create_ids(["pop1", "pop2"], [1, 3])
+            >>> nodes.ids(group=node_ids)  #  returns ID 1 from pop1 and ID 3 from pop2
+            >>> nodes.ids(group=0)  #  returns CircuitNodeIds 0 from pop1 and pop2
+            >>> nodes.ids(group=[0, 1])  #  returns CircuitNodeIds 0 and 1 from pop1 and pop2
+            >>> nodes.ids(group="node_set_name")  # returns CircuitNodeIds matching node set
+            >>> nodes.ids(group={Node.LAYER: 2})  # returns CircuitNodeIds matching layer==2
+            >>> nodes.ids(group={Node.LAYER: [2, 3]})  # returns CircuitNodeIds with layer in [2,3]
+            >>> nodes.ids(group={Node.X: (0, 1)})  # returns CircuitNodeIds with 0 < x < 1
+            >>> # returns CircuitNodeIds matching one of the queries inside the 'or' list
+            >>> nodes.ids(group={'$or': [{ Node.LAYER: [2, 3]},
+            >>>                          { Node.X: (0, 1), Node.MTYPE: 'L1_SLAC' }]})
+            >>> # returns CircuitNodeIds matching all the queries inside the 'and' list
+            >>> nodes.ids(group={'$and': [{ Node.LAYER: [2, 3]},
+            >>>                           { Node.X: (0, 1), Node.MTYPE: 'L1_SLAC' }]})
+        """
+        str_type = "<U{}".format(max(len(pop) for pop in self.population_names))
+        if isinstance(group, CircuitNodeIds):
+            diff = np.setdiff1d(group.get_populations(unique=True), self.population_names)
+            if diff.size != 0:
+                raise BluepySnapError("Population {} does not exist in the circuit.".format(diff))
+        ids = np.array([], dtype=np.int64)
+        populations = np.array([], dtype=str_type)
+        for name, pop in self.items():
+            pop_ids = pop.ids(group=group, raise_missing_property=False)
+            pops = np.array(np.full_like(pop_ids, fill_value=name, dtype=str_type))
+            ids = pop_ids if ids.size == 0 else np.concatenate([ids, pop_ids])
+            populations = pops if populations.size == 0 else np.concatenate([populations, pops])
+        return CircuitNodeIds.create_ids(populations, ids)
+
+    def get(self, group=None, properties=None):
```

Comment:
Am I correct that if some properties are not presented, we will have a BluepySnapError? If yes, should we describe it in `Raises:`?

Reply:
No we don t have a raise. But we should. I will add it. 
Actually the behavior should be : 
if the property is not in the circuit --> raise. 
if the property is just not in a some of the population --> fill with NaN.

I have added the check for existing property in the circuit, tests and the srt/list to the docstring.

Reply:
thank you